mlir/lib/Quantum/Transforms/PropagateSimpleStatesAnalysis.hpp

```diff
@@ -0,0 +1,236 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This algorithm is taken from https://arxiv.org/pdf/2012.07711, figure 5
+
+#pragma once
+
+#include <map>
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/Support/Debug.h"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "Quantum/IR/QuantumOps.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+
+// The six Pauli eigenstates
+enum class QubitState {
+    ZERO,
+    ONE,
+    PLUS,
+    MINUS,
+    LEFT,
+    RIGHT,
+    NOT_A_BASIS,
+};
+
+// {input state : {gate, output state}}
+static std::map<QubitState, std::map<StringRef, QubitState>> QubitTransitions = {
+    {QubitState::ZERO,
+     {
+         {"Hadamard", QubitState::PLUS},
+         {"PauliX", QubitState::ONE},
+         {"PauliY", QubitState::ONE},
+         {"PauliZ", QubitState::ZERO},
+     }},
+
+    {QubitState::ONE,
+     {
+         {"Hadamard", QubitState::MINUS},
+         {"PauliX", QubitState::ZERO},
+         {"PauliY", QubitState::ZERO},
+         {"PauliZ", QubitState::ONE},
+     }},
+
+    {QubitState::PLUS,
+     {
+         {"Hadamard", QubitState::ZERO},
+         {"PauliX", QubitState::PLUS},
+         {"PauliY", QubitState::MINUS},
+         {"PauliZ", QubitState::MINUS},
+         {"S", QubitState::LEFT},
+     }},
+
+    {QubitState::MINUS,
+     {
+         {"Hadamard", QubitState::ONE},
+         {"PauliX", QubitState::MINUS},
+         {"PauliY", QubitState::PLUS},
+         {"PauliZ", QubitState::PLUS},
+         {"S", QubitState::RIGHT},
+     }},
+
+    {QubitState::LEFT,
+     {
+         {"Hadamard", QubitState::RIGHT},
+         {"PauliX", QubitState::RIGHT},
+         {"PauliY", QubitState::LEFT},
+         {"PauliZ", QubitState::RIGHT},
+         // We leave in S+ to indicate the FSM structure
+         // The actual implementation is `quantum.custom "S"() %in {adjoint}`
+         //{"S+", QubitState::PLUS},
+     }},
+
+    {QubitState::RIGHT,
+     {
+         {"Hadamard", QubitState::LEFT},
+         {"PauliX", QubitState::LEFT},
+         {"PauliY", QubitState::RIGHT},
+         {"PauliZ", QubitState::LEFT},
+         // We leave in S+ to indicate the FSM structure
+         // The actual implementation is `quantum.custom "S"() %in {adjoint}`
+         //{"S+", QubitState::MINUS},
+     }},
+};
+
+class PropagateSimpleStatesAnalysis {
+  public:
+    PropagateSimpleStatesAnalysis(Operation *target)
+    {
+        // `target` is a qnode function
+        target->walk([&](Operation *op) {
+            if (op->getNumResults() != 1) {
+                // restrict to single-qubit gates
+                return;
+            }
```

Comment:
What happens if you only have two or more qubit gates?

What happens if there is control flow? I think you are missing this. E.g.,

```python
if path0:
  qml.Hadamard(0)
else: #path1
  qml.PauliX(0)
# state_of_wire_0_bb2 join(state_of_wire_0_bb0, state_of_wire_0_bb1)
# which a safe answer would be NOT_A_BASIS
```

I think this restriction happens because of the line below:

```
            Value res = op->getResult(0);
            if (!isa<quantum::QubitType>(res.getType())) {
                // not a qubit value
                return;
            }
```

But I would make it explicit. I.e., create a function called: `isValidTransformationCandidate` and write an is `isQuantumOp` function.

Reply:
Actually maybe isQuantumOp wouldn't work because you also have to admit either stablehlo constants or arith constants. But maybe isNotInvalidDialect and add SCF there. You should also add CallOp there. Technically, at the moment all quantum operations will be in a single function so this is not necessary, but hopefully in the future we should be able to pass a quantum register / qubits to a function and do quantum operations there as well.